src/main/java/org/wildfly/security/auth/spi/Attributes.java

```diff
@@ -576,13 +598,81 @@ default boolean isEmpty() {
     }
 
     /**
-     * Get the collection of values for the given key.  The result may implement {@link SetEntry} if the values
-     * are distinct (for example, a role or group set).
+     * Returns a read-only instance of this instance.
      *
-     * @param key the attribute name
-     * @return the (possibly empty) attribute value collection
+     * @return a read-only instance of this instance.
      */
-    Entry get(String key);
+    default Attributes asReadOnly() {
+        return new Attributes() {
+            private Map<String, Entry> entryCache;
+            private Collection<Entry> cachedEntries;
+
+            @Override
+            public Collection<Entry> entries() {
+                if (this.cachedEntries != null) {
+                    return this.cachedEntries;
+                }
+
+                return this.cachedEntries = new AbstractCollection<Entry>() {
+                    Map<String, Entry> entryCache = getEntryCache();
+
+                    @Override
+                    public Iterator<Entry> iterator() {
+                        Iterator<Entry> iterator = entryCache.values().iterator();
+
+                        return new Iterator<Entry>() {
+                            @Override
+                            public boolean hasNext() {
+                                return iterator.hasNext();
+                            }
+
+                            @Override
+                            public Entry next() {
+                                return iterator.next();
+                            }
+                        };
+                    }
+
+                    @Override
+                    public int size() {
+                        return this.entryCache.size();
+                    }
+                };
+            }
+
+            @Override
+            public int size(String key) {
+                return Attributes.this.size(key);
+            }
+
+            @Override
+            public String get(String key, int idx) {
+                return Attributes.this.get(key, idx);
+            }
+
+            @Override
+            public int size() {
+                return Attributes.this.size();
+            }
+
+            @Override
+            public Entry get(String key) {
+                return getEntryCache().get(key);
```

Comment:
If the cache doesn't contain that key, you have to create a wrapper entry for the cache.